### DISCUSSION TITLE
Using SSHFS homes and SSSD Active Directory authentication

### DISCUSSION THREAD

**Discussion by vsuojanen** [ltsp/ltsp#516] [General]
There's a large number of examples on the internet how to join Linux clients to the Active Directory domain with SSSD components but it's difficult to find examples or hints how setup and join clients in LTSP network (e.x. in default chrootless system or headless/chroot environment)

I recently joined one LTSP client to AD and after test configuration tried to log in as AD user from GUI and terminal . SSSD provided me the authentication and user information but the login failed in both GUI and terminal. The pamltsp was also giving me  an error `.Pamltsp failed to mount home via SSHFS: read: Connection reset by peer.`  I will continue on this issue more below but that error message is not a reason for failed login. I found it after debugging  the clients syslogs and SSSD logs (reading logs directly from the client). <-- thanks @alkisg. 

f you don't get any shell after login it means there are some problems with the AD and/or SSSD configuration.  For my AD environment I needed to add `ad_gpo_ignore_unreadable = True` in client /etc/sssd/sssd.conf configuration for login to be successful from the LTSP client. 

The error `.Pamltsp failed to mount home via SSHFS: read: Connection reset by peer.` is telling about the SSHFS home directory and that's what is needed after successful login. Below is how I got AD authentication and SSHFS home directories to work on LTSP clients. 

**Do not use the below information in production environment, the purpose is for test environments and for discussion**

I joined the clients to AD domain manually using adcli.
Before joining LTSP clients to AD domain:
- Active Directory is active, managed and maintained by you, AD Administrator or administrator group
- you, Administrator or  the administrator group can join or enroll client machines to the AD 
- the AD domain controller is providing active DNS in the network
- system clocks are right and synced (e.x. using same NTP server)

Install SSSD components and tools for the LTSP server and clients (and  in chroot if using it for client image):

```
sudo apt-get install sssd libpam-sss libnss-sss sssd-tools adcli
sudo apt install krb5-user sssd-krb5
sudo apt install samba-common-bin 
```
(not sure if the samba-common-bin is needed (for sssd or adcli? for `net ads` command at least, see below) 

Do not update the client image yet, there's some configuration and preparation steps for the clients

A summary of SSSD configuration steps:

-	Configuring Kerberos and  KDC server information for both clients and server in /etc/krb5.conf
-	Configuring the hosname of the LTSP server and joining it to the AD
-	Checking keytab exist and is working on the LTSP server
-	Configuring /etc/sssd/sssd.conf on the LTSP server
-	Configuring the hostname of the clients to the /etc/ltsp/ltsp.conf and joining tthem one by one to the AD
-	Collecting keytabs for the clients (e.x. on the LTSP server)
-	Configuring and symlinking a separate /etc/sssd/sssd.conf for the clients /etc/ltsp/sssd/sssd.conf (if need to be different for some client)

With the defaults using SSHFS homes LTSP client boot expects the user to exist in local passwd and group information. The information is collected from the server and exported to clients (local) passwd and group information. If the user enters in to login prompt credentials that match to the exported LTSP server /etc/passwd and /etc/group then pamltsp uses this information in ssh to mount  the home directory (sshfs) from the server. 

When configuring AD users to the clients, using the user id and group information (UID/GID) is no different from the default LTSP with exported passwd and group information. For sshfs home directories the users need to exist both on the LTSP server and on the clients. Configuring AD users in LTSP network with sshfs homes you need to join both server and clients for sshfs home directories to work.

### Setup Kerberos and Samba
(for server and client)
Configure /etc/krb5.conf and /etc/samba/smb.conf as below changing the domains suitable to your own environment:

/etc/krb5.conf
```

 [logging]
 default = FILE:/var/log/krb5libs.log

 [libdefaults]
 default_realm = AD.EXAMPLE.COM
 dns_lookup_realm = true
 dns_lookup_kdc = true
 ticket_lifetime = 24h
 renew_lifetime = 7d
 forwardable = true
 rdns = false

 # You may also want either of:
 # allow_weak_crypto = true
 # default_tkt_enctypes = arcfour-hmac

 [realms]
 # Define only if DNS lookups are not working
 # AD.EXAMPLE.COM = {
 # kdc = server.ad.example.com
 # master_kdc = server.ad.example.com
 # admin_server = server.ad.example.com
 # }

 [domain_realm]
 # Define only if DNS lookups are not working
 # .ad.example.com = AD.EXAMPLE.COM
 # ad.example.com = AD.EXAMPLE.COM
```

/etc/samba/smb.conf

```
 [global]
 security = ads
 realm = AD.EXAMPLE.COM
 workgroup = EXAMPLE

 log file = /var/log/samba/%m.log

 kerberos method = secrets and keytab

 client signing = yes
 client use spnego = yes
```

This ref. to the official SSSD instructions from RedHat:
https://sssd.io/docs/ad/ad-provider-manual.html

### Joining to AD
If you have rights to join machines to AD domain then join first the LTSP server and  then clients using `net ads` or `adcli`. If you do not have rights to join machines to domain ask your Windows AD server administrator or domain administrators to enroll the hostnames in AD and transfer the keytabs.

From the server and clients create a Kerberos ticket  with username having rights to join machines to  the domain:
kinit user@domain

Join the machine with samba/Windows net logon (net ads):
`net ads join -k `
or 
`net ads join -U user@domain`

This needs to be done from both server and clients  with correctly enrolled and configured hostnames (e.x. host.example.com). 
With the above net command the client keytab gets created in machines /etc/krb5.keytab. 

For the server leave it as it is created in the server file etc/krb5.keytab

For the clients collect you need to collect keytabs after creating from each client /etc/krb5.keytab. Collect the client keytabs to server as /etc/krb5.keytab.hostname

Hint: Using adcli you can join the clients from server to AD domain giving the client machines hostname/remote machine hostname and directing the keytab to separate files).

e.x. `adcli join -U USER@DOMAIN -H host.example.com -K /etc/krb5.keytab.hostname example.com`
(This avoids transferring the keytabs between hosts but also booting each client)

Then after having all keytabs in place on the server make sure rights are correct:
```
chown root:root /etc/krb5.keytab
chmod 0600 /etc/krb5.keytab

chown root:root /etc/krb5.keytab.hostname
chmod 0600 /etc/krb5.keytab.hostname

etc.
```

Add the following in ltsp.conf under each client:

/etc/ltsp/ltsp.conf
```
...

[00:16:3e:88:10:b2]
HOSTNAME=hostname.example.com
POST_INIT_LN_KRB5_HOSTKEY="ln -sf /etc/krb5.keytab.hostname /etc/krb5.keytab"
```

Finally configure the server and clients to use the SSSD modules and Active Directory for authentication and login

/etc/sssd/sssd.conf 

```
[sssd]
config_file_version = 2
domains = example.com
services = nss, pam

[domain/example.com]
# debug_level=9
# Uncomment if you need offline logins
# cache_credentials = true

id_provider = ad
auth_provider = ad
access_provider = ad
# Uncomment for debugging e.x. access/permissions
#access_provider = permit

# Uncomment if service discovery is not working
# ad_server = server.ad.example.com

# Uncomment if you want to use POSIX UIDs and GIDs set on the AD side
# ldap_id_mapping = False

# Uncomment if the trusted domains are not reachable
#ad_enabled_domains = ad.example.com

# Comment out if the users have the shell and home dir set on the AD side
default_shell = /bin/bash
fallback_homedir = /home/%d/%u

# Uncomment and adjust if the default principal SHORTNAME$@REALM is not available
# ldap_sasl_authid = host/client.ad.example.com@AD.EXAMPLE.COM

# Comment out if you prefer to use shortnames.
use_fully_qualified_names = True

# Uncomment if the child domain is reachable, but only using a specific DC
# [domain/ad.example.com/child.example.com]
# ad_server = dc.child.example.com

# Uncomment if Group Policy is unreadable or has unreadable or missing attribute problems 
# ad_gpo_ignore_unreadable = True
# DNS problems?
# ad_gpo_access_control = permissive


```
[Ubuntu SSSD instructions ](https://ubuntu.com/server/docs/service-sssd) tell that you need to enable separately mkhomedir module (pam_auth ?). It didn't break anything in my tests when I enabled it:

sudo pam-auth-update --enable mkhomedir

Test the configuration works first on the server rebooting the server and login with any AD user.

Login to your server as aduser@example.com
Test also login from the server using ssh aduser@example.com@ipaddress_of_server

Now if everything works on the server update the client image with the configuration above (check for  symlinks to client keytabs, and /etc/krb5.conf, /etc/samba/smb.conf, /etc/sssd/sssd.conf )

Test the client

I apologize if the above information is not working for you. 
It's for discussion based on my first test with sssd + ad + sshfs where I got the LTSP client login and home directory successful.

Comments are welcome

Related:
https://ubuntu.com/server/docs/service-sssd

**Comment by Soylor** [comment]
I got it working but with this modification to **sssd.conf**:
`ad_gpo_access_control = disabled`

**Comment by niansa** [comment]
I joined the domain via realmd: https://access.redhat.com/documentation/de-de/red_hat_enterprise_linux/7/html/windows_integration_guide/ch-configuring_authentication
It did set up automatic domain joining for clients too (they can access the domain), however, sshfs home is broken for me too.

**Reply by viniciusferrao** [reply]
How did you setup automatic joining for clients with `realmd`?

**Comment by viniciusferrao** [comment]
Is there any fix to the `.Pamltsp failed to mount home via SSHFS:.` message that shows up on `lightdm`?

**Reply by vsuojanen** [reply]
This Post here is originally to share all that information i got that Time to resolve and configure that sshfs home and get the error message away. Do you manage to get to the desktop after login despite The lightdm error message?

**Reply by viniciusferrao** [reply]
I've ended up enabling NFS for home directories, way cleaner and faster. I didn't get the security benefit of using SSHFS instead.

**Comment by viniciusferrao** [comment]
Still on this matter, how are you guys dealing with `krb5.keytab` renewals? Here after 30 days the system just don't renews the keytab and no one is able to login again. I have to re-enroll the clients every 30 days.

**Reply by vsuojanen** [reply]
Are the AD user id's (e.x. your own id) able to login on the server after the 30 days?

On the server I believe the ticket should be created/renewed automatically when any AD user is logging in locally e.x. from display manager or terminal (on the LTSP 'server' the file is etc/krb5.keytab). But the server is not of course the machine where the user sits.

The worst thing here is that when client boots up from LTSP server you need to have the right client machine file in place in client /etc/krb5.keytab. In the original post for the LTSP clients hostnames I didn't figure out anything more useful to workaround with the issue but the symlink trick and then schedule for the extra maintenance. 

e.x. adcli join -U USER@DOMAIN -H host.example.com -K /etc/krb5.keytab.hostname example.com

Then this corresponding ltsp.conf configuration for each client machine: 

[00:16:3e:88:10:b2]
HOSTNAME=hostname.example.com
POST_INIT_LN_KRB5_HOSTKEY="ln -sf /etc/krb5.keytab.hostname /etc/krb5.keytab"

**Reply by viniciusferrao** [reply]
On the server everything is fine, the problem is inside the chroot, since I use a separate chroot. To be honest I've took a shortcut and joined the image directly, so every client uses the same Kerberos Ticket that expires on 30 days and on AD I have a single computer account. Security is not a big deal with my LTSP environment.

Where do you run the `adcli` command? On the server? Is it on `crontab`? Because that's the missing part that I'm not understanding correctly.
